### PR TITLE
Add database token request for twitter api.py #98

### LINKED ISSUE #98: Add database token request for twitterApi.py
## Description
`twitterApi.py` is still using a fixed token and needs to use the neo4j tokens to not run out of requests

### PR DESCRIPTION
**Pull request by Eik-S**
Resolves #98

~~**Don't merge before #95, #99**~~

# Important changes

- rewrote twitterApi as a class
- added tokenProvider class to care for the token switching

# Review hints
Already deployed

Just use the tool and watch if there are any errors thrown in the api.
You could also dive into the code because i'm not super confident with object oriented python (if we have time for that atm)

### CODE REVIEW

File: api/twitter/tokenProvider.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                    if self.twitter_route in route:
+                        reset_time = parsed_response[category][route]['reset']
+            query = "MATCH (h:TOKEN{token:'%s'}) " % self.user_token
+            query += "SET h.`%s`=%s" % (self.twitter_route, reset_time)
+            self.__run_query(query)
+        except Exception as exc:
+            print("14 - ERROR -> %s. Failed to reset the token's timestamp." % exc)
+            if 'errors' in parsed_response:
+                e_code = parsed_response['errors'][0]['code']
+                if e_code == 89:
```
Comment (jonathanschlue) [inline api/twitter/tokenProvider.py]: At some point, we should have a concise mapping from twitter error codes to corresponding error message. We could maintain them in a file `constants.py`.

File: api/twitter/twitterApi.py [hunk 2]
```diff
@@ -2,168 +2,210 @@
...
+                return parsed_response
+
+    def __check_error(self, api, response):
+        error_response = ""
+        if 'error' in response:
+            error_response = self.__check_twitter_error_code(
+                response["error"])
+        elif 'errors' in response:
+            error_response = self.__check_twitter_error_code(
+                response["errors"][0]["code"])
```
Comment (jonathanschlue) [inline api/twitter/twitterApi.py]: What about the case that there is more than one error in `response["errors"]`?

File: api/twitter/twitterApi.py [hunk 3]
```diff
@@ -2,168 +2,210 @@
...
+        if error_response == "":
+            return False
+        else:
+            if error_response == "Switch helper":
+                api.__get_user_auth()
+                return "continue"
+            elif error_response == "Invalid user":
+                return "invalid user"
+            elif error_response == "Not authorized":
+                return "invalid user"
```
Comment (jonathanschlue) [inline api/twitter/twitterApi.py]: I suggest

```python
                elif error_response in ("Invalid user", "Not authorized"):
                    return "invalid user"
```

File: api/twitter/twitterApi.py [hunk 4]
```diff
@@ -2,168 +2,210 @@
...
+
+    def get_retweeters(self, tweet_id):
+        """Request the 100 last retweet ids, return them as a list"""
+        route = 'statuses/retweeters/ids'
+        params = { 'id': str(tweet_id)}
+        data = self.__request_twitter(route, params)
+        response = {}
+        response['response'] = data['ids']
+        retweeters = response['response']
+        """change user_ids from num to string"""
```
Comment (jonathanschlue) [inline api/twitter/twitterApi.py]: doc-strings shouldn't be used inside function bodies except for the very first line.

```suggestion
        # change user_ids from num to string
```

File: api/twitter/twitterApi.py [hunk 5]
```diff
@@ -2,168 +2,210 @@
...
+        response['response'] = {}
+        response['response'][ data['id_str']] = {}
+        tweet_dict = response['response'][ data['id_str']]
+        tweet_dict["reply_to"] = str( data['in_reply_to_status_id_str'])
+        tweet_dict["lang"] = str( data['lang'])
+        tweet_dict["author"] = str( data['user']['id_str'])
+        tweet_dict["fav_count"] = str( data['favorite_count'])
+        tweet_dict["retweet_count"] = str( data['retweet_count'])
+        tweet_dict["date"] = str( data['created_at'])
+        """The following values are lists"""
```
Comment (jonathanschlue) [inline api/twitter/twitterApi.py]:
```suggestion
     # The following values are lists
```

File: api/twitter/twitterApi.py [hunk 6]
```diff
@@ -2,168 +2,210 @@
...
+        }
+        route = "statuses/user_timeline"
+        data = self.__request_twitter(route, params)
+        return data
+
+    @staticmethod
+    def __parse_properties( data, keys):
+        response = {}
+        for key in keys:
+            response[key] = data[key]
```
Comment (jonathanschlue) [inline api/twitter/twitterApi.py]: beware of `KeyError`s here.

File: api/twitter/tokenProvider.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                print("0 - ERROR -> %s. Crawler could not connect to the database. Retrying in 5s..." % exc)
+                time.sleep(5)
+                continue
+
+    def __get_user_auth(self):
+        """
+        Return old token by updating the reset time
+        and get a free token.
+        """
+        if hasattr(self, 'twitter_api'):
```
[the remark this answers is not in the document]
Comment (Eik-S) [reply api/twitter/tokenProvider.py]: has been a mistake, changed to `'api'`

File: api/twitter/tokenProvider.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        while True:
+            try:
+                driver = GraphDatabase.driver(
+                    os.environ.get('NEO4J_URI'), auth=(
+                        os.environ.get('NEO4J_USER'),
+                        os.environ.get('NEO4J_PASSWORD')
+                    )
+                )
+                return driver
+            except Exception as exc:
```
[the remark this answers is not in the document]
Comment (Eik-S) [reply api/twitter/tokenProvider.py]: I would suggest to leave it as it is for now.
I created #104 to add python3 logging and exchange all prints through meaningful logging messages. With doing so i would use 'error' level logs for generic exceptions.
After logging those exceptions we know more easy which exceptions are occuring when and can add specific catches for them.

File: api/twitter/tokenProvider.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            query += "WHERE NOT EXISTS(h.`%s`) " % self.twitter_route
+            query += "OR h.`%s` < %s " % (self.twitter_route, actual_time)
+            query += "WITH h LIMIT 1 "
+            query += "SET h.`%s`=%s " % (self.twitter_route, distant_time)
+            query += "RETURN h.token as token, "
+            query += "h.secret as secret, "
+            query += "h.user as user"
+            with self.neo4j_driver.session() as db:
+                results = db.run(query)
+                for user in results:
```
[the remark this answers is not in the document]
Comment (Eik-S) [reply api/twitter/tokenProvider.py]: There is just one, copied most of the code from the crawler and didn't check this

File: api/twitter/twitterApi.py (hunk 2 above)
Comment (Eik-S) [reply api/twitter/twitterApi.py]: This should also be part of #104

File: api/twitter/twitterApi.py (hunk 3 above)
Comment (Eik-S) [reply api/twitter/twitterApi.py]: :+1:

File: api/twitter/tokenProvider.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jonathanschlue) [reply api/twitter/tokenProvider.py]: I aggree, lets then do it like that :)

File: api/twitter/twitterApi.py (hunk 2 above)
Comment (jonathanschlue) [reply api/twitter/twitterApi.py]: Despite of using print or logging.err, is twitter response sometimes containing more than one error?

**Comment by Eik-S** [comment]
My vscode is somehow not able to see autopep8 neither can install it... 
Dont want to fix that know with switching my laptop/os tomorrow... :grin: 

Tried to resolve everything else as good as i can.
Some error refinements are task of #104

File: api/twitter/tokenProvider.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jonathanschlue) [reply api/twitter/tokenProvider.py]: k

**Comment by jonathanschlue** [comment]
Merged into #105 according to @Eik-S in [this comment](https://github.com/tracemap/tracemap-backend/issues/98#issuecomment-437653275)

**Pull request closed without merging**